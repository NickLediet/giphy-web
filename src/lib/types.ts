
export enum TextPosition {
    CenterTop = 'center-top',
    CenterBottom = 'center-bottom',
    Below = 'below',
}

export type Image = {
    url: string,
    altText: string,
    id: string
}