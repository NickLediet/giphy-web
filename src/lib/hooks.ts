import { useState } from 'react'
import { Image } from './types'
import { memoize } from 'lodash-es'
const API_KEY = import.meta.env.VITE_GIPHY_API_KEY
const REQUEST_LIMIT = 3
const REQUEST_RATING = 'g'


export function useImageQuery() {
    const [images, setImages] = useState<Image[]>([])
    const [error, setError] = useState<string>()

    const queryImages = memoize(async function _queryImages(query: string) {
        try {
            const { data } = await fetch(`https://api.giphy.com/v1/stickers/search?q=${query}&limit=${REQUEST_LIMIT}&rating=${REQUEST_RATING}&api_key=${API_KEY}`)
                .then(res => res.json())
            console.log(data)
            setImages(data.map((image: any) => ({
                id: image.id,
                url: image.images.downsized_medium.url,
                altText: image.title
            })))
        } catch(e) {
            if (typeof e === "string") {
                setError(e.toUpperCase())
            } else if (e instanceof Error) {
                setError(e.message) // works, `e` narrowed to Error
            }
        }
    })

    return {
        images, error, queryImages
    }

}