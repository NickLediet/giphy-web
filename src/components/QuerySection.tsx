import {Button, TextField, Select } from '@radix-ui/themes'
import {MagnifyingGlassIcon} from '@radix-ui/react-icons'
import {useState} from 'react'
import { TextPosition } from '../lib'

type QuerySectionProps = {
    onQueryChange?: (query: string, position: TextPosition, text: string) => void
}

export function QuerySection({ onQueryChange }: QuerySectionProps) {
    const [query, setQuery] = useState('')
    const [position, setPosition] = useState<TextPosition>()
    const [text, setText] = useState('')

    function submitHandler(event: React.FormEvent) {
        event.preventDefault()
        onQueryChange && onQueryChange(query, position || TextPosition.Below, text)
    }

    return (
        <form onSubmit={submitHandler}>
            <TextField.Root 
                data-testid="query"
                placeholder="Search for images..."
                onInput={(event) => setQuery(event.currentTarget.value)}
                value={query}>
                <TextField.Slot>
                    <MagnifyingGlassIcon height="16" width="16" />
                </TextField.Slot>
            </TextField.Root>

            <Select.Root 
                data-testid="position"
                value={position} 
                onValueChange={(value) => setPosition(value as TextPosition)}>
                <Select.Trigger placeholder="Select text position"/>
                <Select.Content>
                    <Select.Group>
                        <Select.Label>Text Position</Select.Label>
                        <Select.Item value={TextPosition.CenterTop}>Center Top</Select.Item>
                        <Select.Item value={TextPosition.CenterBottom}>Center Bottom</Select.Item>
                        <Select.Item value={TextPosition.Below}>Below Image</Select.Item>
                    </Select.Group>
                </Select.Content>
            </Select.Root>

            <TextField.Root
                data-testid="text"
                placeholder="Search the docsâ€¦" 
                onInput={(event) => setText(event.currentTarget.value)}
                value={text} />

            <Button type="submit" data-testid="submit">Submit</Button>
        </form>
    )
}