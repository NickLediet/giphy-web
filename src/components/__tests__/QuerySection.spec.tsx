import { describe, expect, it, vi } from 'vitest'
import { fireEvent, render, screen } from '@testing-library/react'
import { Theme } from '@radix-ui/themes'
import { TextPosition } from '../../lib'
import { QuerySection } from '../QuerySection'

window.ResizeObserver = vi.fn().mockImplementation(() => ({
    observe: vi.fn(),
    unobserve: vi.fn(),
    disconnect: vi.fn(),
}))
describe('QuerySection', () => {    
    it('matches snapshot', () => {
        const { container, unmount } = render(<Theme>
            <QuerySection onQueryChange={() => {}} />
        </Theme>)
        expect(container).toMatchSnapshot()
        unmount()
    })

    it('calls onQueryChange when query changes', () => {
        const onQueryChange = vi.fn((query: string, position: TextPosition, text: string) => {
            console.log(query, position, text)
        })
        const {unmount, container} = render(<Theme>
            <QuerySection onQueryChange={onQueryChange} />
        </Theme>)
        // Update query input
        const queryInput = screen.getByTestId('query') as HTMLInputElement
        const query = 'test query'
        fireEvent.input(queryInput, { target: { value: query } })
        // Update Position select
        const positionSelect = container.querySelector('select') as HTMLSelectElement
        const position = TextPosition.CenterTop
        fireEvent.change(positionSelect, { target: { value: position } })
        // Update text input
        const textInput = screen.getByTestId('text') as HTMLInputElement
        const text = 'test text'
        fireEvent.input(textInput, { target: { value: text } })
        // Submit form
        fireEvent.submit(
            container.querySelector('form') as HTMLFormElement,
        )

        expect(onQueryChange).toHaveBeenCalledWith(query, position, text)
        unmount()
    })
})