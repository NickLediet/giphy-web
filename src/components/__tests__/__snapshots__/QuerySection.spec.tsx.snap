// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`QuerySection > matches snapshot 1`] = `
<div>
  <div
    class="radix-themes"
    data-accent-color="indigo"
    data-gray-color="slate"
    data-has-background="true"
    data-is-root-theme="true"
    data-panel-background="translucent"
    data-radius="medium"
    data-scaling="100%"
  >
    <form
      class="query-section-container"
    >
      <h2
        class="rt-Heading rt-r-size-4 rt-r-ta-left rt-r-mb-3"
      >
        Search for images
      </h2>
      <div
        class="rt-Grid rt-r-gtc-1 sm:rt-r-gtc-2 rt-r-gap-3 rt-r-w"
        style="--width: auto;"
      >
        <div
          class="rt-Box rt-r-gcs rt-r-gce"
          style="--grid-column-start: 1; --grid-column-end: -1;"
        >
          <div
            class="rt-TextFieldRoot rt-r-size-2 rt-variant-surface"
            data-accent-color=""
          >
            <input
              class="rt-reset rt-TextFieldInput"
              data-testid="query"
              placeholder="Search for images..."
              spellcheck="false"
              value=""
            />
            <div
              class="rt-TextFieldSlot"
            >
              <svg
                fill="none"
                height="16"
                viewBox="0 0 15 15"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  clip-rule="evenodd"
                  d="M10 6.5C10 8.433 8.433 10 6.5 10C4.567 10 3 8.433 3 6.5C3 4.567 4.567 3 6.5 3C8.433 3 10 4.567 10 6.5ZM9.30884 10.0159C8.53901 10.6318 7.56251 11 6.5 11C4.01472 11 2 8.98528 2 6.5C2 4.01472 4.01472 2 6.5 2C8.98528 2 11 4.01472 11 6.5C11 7.56251 10.6318 8.53901 10.0159 9.30884L12.8536 12.1464C13.0488 12.3417 13.0488 12.6583 12.8536 12.8536C12.6583 13.0488 12.3417 13.0488 12.1464 12.8536L9.30884 10.0159Z"
                  fill="currentColor"
                  fill-rule="evenodd"
                />
              </svg>
            </div>
          </div>
        </div>
        <div
          class="rt-Box rt-r-gcs rt-r-gce"
          style="--grid-column-start: 1; --grid-column-end: -1;"
        >
          <div
            class="rt-TextFieldRoot rt-r-size-2 rt-variant-surface"
            data-accent-color=""
          >
            <input
              class="rt-reset rt-TextFieldInput"
              data-testid="text"
              placeholder="Image text..."
              spellcheck="false"
              value=""
            />
          </div>
        </div>
        <button
          aria-autocomplete="none"
          aria-controls="radix-:r0:"
          aria-expanded="false"
          class="rt-reset rt-SelectTrigger rt-r-size-2 rt-variant-surface"
          data-accent-color=""
          data-placeholder=""
          data-state="closed"
          dir="ltr"
          role="combobox"
          type="button"
        >
          <span
            class="rt-SelectTriggerInner"
          >
            <span
              style="pointer-events: none;"
            >
              Select text position
            </span>
          </span>
          <svg
            aria-hidden="true"
            class="rt-SelectIcon"
            fill="currentcolor"
            height="9"
            viewBox="0 0 9 9"
            width="9"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0.135232 3.15803C0.324102 2.95657 0.640521 2.94637 0.841971 3.13523L4.5 6.56464L8.158 3.13523C8.3595 2.94637 8.6759 2.95657 8.8648 3.15803C9.0536 3.35949 9.0434 3.67591 8.842 3.86477L4.84197 7.6148C4.64964 7.7951 4.35036 7.7951 4.15803 7.6148L0.158031 3.86477C-0.0434285 3.67591 -0.0536285 3.35949 0.135232 3.15803Z"
            />
          </svg>
        </button>
        <select
          aria-hidden="true"
          style="position: absolute; border: 0px; width: 1px; height: 1px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; word-wrap: normal;"
          tabindex="-1"
        >
          <option
            value=""
          />
          <option
            value="center-top"
          >
            Center Top
          </option>
          <option
            value="center-bottom"
          >
            Center Bottom
          </option>
          <option
            value="below"
          >
            Below Image
          </option>
        </select>
        <button
          class="rt-reset rt-BaseButton rt-r-size-2 rt-variant-solid rt-Button"
          data-accent-color=""
          data-testid="submit"
          type="submit"
        >
          Submit
        </button>
      </div>
    </form>
  </div>
</div>
`;
